### PR TITLE
[type] Add BitArrayType and corresponding SNodes

### PR DESCRIPTION
**Pull request by TH3CHARLie**
Related issue = #1905, using #1968 as reference

- Added `BitArrayType`
- Added corresponding SNode API
- `misc/test_bit_array.py` for demonstrating snode structures(and also a testcase for the future)

I am not sure if the syntax in `misc/test_bit_array.py` is ideal for constructing a bit array, which mainly mimics the bit_struct behavior so the struct compiler can get to know the `num_element` field by counting the number of children

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

**Review by yuanming-hu — COMMENTED** [review]
Cool! I think it's a nice start. Regarding the construction API: I think we would need something like `ti.root._bit_array(ti.i, 32, num_bits=32).place(x1)` to use 32 out of 32 bits. 

The `bit_array` SNodes should share a similar structure to `dense`. They are bot arrays, but `bit_array` has a finer granularity.

> the struct compiler can get to know the num_element field by counting the number of children

As noted above, we want the user to specify the number of elements there :-) For now, you can assume there is at most one child of each `bit_array` SNode.

**Review by Hanke98 — APPROVED** [review]
Great job! LGTM. Let's move on to implement CustomInt SNode read/write, so we can do more test here.

**Comment by Hanke98** [comment]
> Cool! I think it's a nice start. Regarding the construction API: I think we would need something like `ti.root._bit_array(ti.i, 32, num_bits=32).place(x1)` to use 32 out of 32 bits.
> 
> The `bit_array` SNodes should share a similar structure to `dense`. They are bot arrays, but `bit_array` has a finer granularity.

I didn't realize the real way we construct the `bit_array` should be like this and feel so sorry for approving this pr for now. I will do it more carefully next time.

**Comment by TH3CHARLie** [comment]
Thanks for the initial review @yuanming-hu @Hanke98! I've updated the bit_array SNode construction API according to yuanming's suggestion.

File: taichi/struct/struct_llvm.cpp [hunk 1]
```diff
@@ -73,6 +73,16 @@ void StructCompilerLLVM::generate_types(SNode &snode) {
     snode.dt = TypeFactory::get_instance().get_bit_struct_type(
         snode.physical_type, ch_types, ch_offsets);
 
+    DataType container_primitive_type(snode.physical_type);
+    body_type = tlctx->get_data_type(container_primitive_type);
+  } else if (type == SNodeType::bit_array) {
+    // bit array snode should only have at one children(element)
```
Comment (yuanming-hu) [inline taichi/struct/struct_llvm.cpp:79]:
```suggestion
    // A bit array SNode should only have one child
```

**Review by yuanming-hu — APPROVED** [review]
Thank you so much! LGTM.

**Review by Hanke98 — APPROVED** [review]
Great! Learnt a lot from your implementation and discussions! Thanks!

**Comment by TH3CHARLie** [comment]
I'll merge this if no one against it.

**Pull request merged**